posthog/temporal/utils.py

```diff
@@ -1,35 +1,116 @@
-    @wraps(fn)
-    async def wrapper(*args, **kwargs):
-        heartbeat_timeout = activity.info().heartbeat_timeout
-        heartbeat_task = None
-        if heartbeat_timeout:
-            # Heartbeat twice as often as the timeout
-            heartbeat_task = asyncio.create_task(heartbeat_every(heartbeat_timeout.total_seconds() / 2))
         try:
-            return await fn(*args, **kwargs)
-        finally:
-            if heartbeat_task:
-                heartbeat_task.cancel()
-                # Wait for heartbeat cancellation to complete
-                await asyncio.wait([heartbeat_task])
+            last_inserted_at = dt.datetime.fromisoformat(details[0])
+        except (TypeError, ValueError) as e:
+            raise HeartbeatParseError("last_inserted_at") from e
+
+        return cls(last_inserted_at, _remaining=details[1:])
+
+
+HeartbeatType = typing.TypeVar("HeartbeatType", bound=HeartbeatDetails)
+
+
+async def should_resume_from_activity_heartbeat(
+    activity, heartbeat_type: typing.Type[HeartbeatType], logger
+) -> tuple[bool, HeartbeatType | None]:
+    """Check if a batch export should resume from an activity's heartbeat details.
+
+    We understand that a batch export should resume any time that we receive heartbeat details and
+    those details can be correctly parsed. However, the decision is ultimately up  to the batch export
+    activity to decide if it must resume and how to do so.
+
+    Returns:
+        A tuple with the first element indicating if the batch export should resume. If the first element
+        is True, the second tuple element will be the heartbeat details themselves, otherwise None.
+    """
+    try:
+        heartbeat_details = heartbeat_type.from_activity(activity)
+
+    except EmptyHeartbeatError:
+        # We don't log this as a warning/error because it's the expected exception when heartbeat is empty.
+        heartbeat_details = None
+        received = False
+        logger.debug("Did not receive details from previous activity execution")
+
+    except NotEnoughHeartbeatValuesError:
+        # This is not an error as it's the expected exception when heartbeat is empty.
+        heartbeat_details = None
+        received = False
+        logger.debug("Did not receive details from previous activity execution")
+
+    except HeartbeatParseError:
+        heartbeat_details = None
+        received = False
+        logger.warning("Details from previous activity execution could not be parsed.")
 
-    return cast(F, wrapper)
+    except Exception:
+        # We should start from the beginning, but we make a point to log unexpected errors.
+        # Ideally, any new exceptions should be added to the previous blocks after the first time and we will never land here.
+        heartbeat_details = None
+        received = False
+        logger.warning(f"Did not receive details from previous activity Excecution due to an unexpected error")
```

Comment:
Should this include the exception itself?

Reply:
Ah, yeah, this log should be promoted to ERROR level and use `logger.exception`.